examples/99_in_tool_selection/ex_in_tool_selection/main.rb

```diff
@@ -0,0 +1,364 @@
+# Licensed under the MIT license
+
+require 'ex_in_tool_selection/inference_lock_helper'
+
+module Examples
+  module InToolSelection
+
+    class ComponentMoveTool
+
+      # Threshold in logical screen pixels for when the mouse is considered to
+      # be dragged.
+      DRAG_THRESHOLD = 10
+
+      def activate
+        @mouse_ip = Sketchup::InputPoint.new
+        @picked_first_ip = Sketchup::InputPoint.new
+        @mouse_position = ORIGIN
+        @mouse_down = ORIGIN
+        @distance = 0
+
+        # Create a InferenceLockHelper object to use in this tool.
+        # See inference_lock_helper.rb for details.
+        @inference_lock_helper = InferenceLockHelper.new
+
+        # Empty the selection if the tool can't use it.
+        # Otherwise keep this preselection.
+        Sketchup.active_model.selection.clear unless valid_selection?
+
+        # The tool selection is tracked separately from the SketchUp selection.
+        # When we are in the selection stage we want to temporarily select
+        # hovered entities for previewing them as selected, but our tool logic
+        # treats the selection as empty until the user clicks.
+        #
+        # Native tools can preview entities as selected without actually
+        # adding them to the selection, but the Ruby API doesn't support this.
+        @selection = Sketchup.active_model.selection.to_a
+
+        # Remember whether tool was activated with preselection or not as this
+        # affect how the tool resets.
+        @preselection = !Sketchup.active_model.selection.empty?
+
+        # The entity being hovered when at the selection stage.
+        @hovered_entity = nil
+
+        @valid_hovered_entity = false
+
+        update_ui
+      end
+
+      def deactivate(view)
+        @inference_lock_helper.unlock
+        view.model.abort_operation if @in_operation
+        view.invalidate
+      end
+
+      def enableVCB?
+        picked_first_point?
+      end
+
+      def resume(view)
+        update_ui
+        view.invalidate
+      end
+
+      def suspend(view)
+        view.invalidate
+      end
+
+      def onCancel(reason, view)
+        reset_tool
+        view.invalidate
+      end
+
+      def onKeyDown(key, _repeat, _flags, view)
+        @inference_lock_helper.on_keydown(key, view, @mouse_ip, @picked_first_ip)
+        pick_mouse_position(view)
+      end
+
+      def onKeyUp(key, _repeat, _flags, view)
+        @inference_lock_helper.on_keyup(key, view)
+        pick_mouse_position(view)
+      end
+
+      def onMouseMove(flags, x, y, view)
+        if @selection.empty?
+          # If the tool selection is empty, try picking the hovered object.
+          try_select_entity(view, x, y)
+        end
+
+        @mouse_position = Geom::Point3d.new(x, y, 0)
+        pick_mouse_position(view)
+      end
+
+      def onLButtonDown(flags, x, y, view)
+        @mouse_down = Geom::Point3d.new(x, y)
+
+        # If tool selection is empty, pick our hovered entity.
```

Comment:
This comment is confusing me. We aren't picking at all at this if branch. Can you clarify?